### ISSUE TITLE
Feature: Model Decomposition

### ISSUE THREAD

**Issue by markvilar** [markvilar/viper#6]
## Problem and Motivation

Pretrained VPR models such as MegaLoc produce very high-dimensional descriptors — MegaLoc emits 8448D vectors from a final linear projection of roughly `(8448, 16640)`. High dimensionality maximises recall on benchmark datasets, but it is a poor fit for the downstream applications viper targets, for two distinct reasons.

**Probabilistic modelling degrades in high dimensions.** Methods such as KappaPlace model uncertainty over descriptor *directions* using the von Mises–Fisher distribution. As dimensionality grows, the hypersphere becomes increasingly sparse, distances concentrate, and concentration (κ) estimates lose discriminative power. Fitting hyperspherical probabilistic models over 8448D descriptors is therefore both statistically fragile and expensive; a lower-dimensional descriptor space is far better conditioned for these methods.

**The final projection dominates the model's memory footprint.** Because it maps ~16640D features to 8448D, MegaLoc's output projection alone accounts for nearly half of the model's weight memory. Shrinking the output dimension `k` cuts this layer roughly proportionally, giving a substantial reduction in the deployed model's size — valuable on memory-constrained platforms independent of any probabilistic use case.

Both problems point to the same need: a principled way to map pretrained descriptors into a lower-dimensional space that preserves as much of the original model's representational structure as possible, rather than discarding information arbitrarily or re-training the projection from scratch. This motivates a reusable model-decomposition capability — SVD-based truncation of an existing linear output layer — that can recompose models like MegaLoc to emit compact descriptors while staying geometrically close to the original at initialisation. Issue #11 depends directly on this capability to produce SVD-truncated MegaLoc checkpoints.

## Proposed Solution

Add a `viper.model_decomposition` sub-package of **generic, model-agnostic** functions for decomposing and remapping the building blocks of pretrained embedders — operating on raw tensors and `nn.Module`s (e.g. `nn.Linear`), with no knowledge of any specific model. This is the reusable foundation; the model-specific *builders* that apply these primitives to a given architecture (a "forge" pattern — e.g. an SVD-truncated MegaLoc builder) are added per model in their own issues. #11 is the first consumer.

The initial primitive is SVD-based truncation of a linear projection, provided at two levels.

**Tensor level** — the mathematical core, decoupled from `nn`:

```python
def truncate_linear_svd(
    weight: Tensor,
    bias: Tensor | None,
    k: int,
) -> tuple[Tensor, Tensor | None]:
    """
    Truncate a linear projection to its k highest-variance output directions
    via truncated SVD.

    Given weight W (shape (out, in)) with W = U Σ Vᵀ, returns the rank-k
    projection into the top-k left-singular (output) subspace:

        W_new = Σ[:k] * Vᵀ[:k, :]     shape (k, in)
        b_new = U[:, :k]ᵀ @ b          shape (k,)

    Arguments:
        weight - projection weight, shape (out, in)
        bias   - projection bias, shape (out,), or None
        k      - target output dimension, 0 < k <= out
    Returns:
        (W_new, b_new) - truncated weight (k, in) and bias (k,) or None
    """
```

**Module level** — a thin convenience returning a ready-to-use layer:

```python
def truncate_linear(linear: nn.Linear, k: int) -> nn.Linear:
    """
    Return a new nn.Linear(in_features, k) initialised from the truncated SVD
    of `linear` (see truncate_linear_svd), mapping into the top-k output
    subspace of the original layer.
    """
```

Truncation keeps the `k` highest-variance output directions; the result is an orthonormal (`U[:, :k]ᵀ`) re-expression of the original descriptor's dominant subspace, so after L2-normalisation descriptors at initialisation stay geometrically close to the original model's, avoiding a cold start during any later fine-tuning.

The sub-package is intended to grow into the home for further generic decomposition/remapping primitives (e.g. PCA/whitening, random projection) as other models need them, keeping model-specific assembly out of the primitives themselves.

**Out of scope for this issue:** the MegaLoc-specific SVD-truncated builder (#11) and any model-registry changes needed to expose truncated variants by key (#15).

## Open questions

None outstanding — see Notes for the settled implementation decisions.

## Notes

- **SVD backend:** use the exact `torch.linalg.svd(full_matrices=False)` (sufficient for the top-`k` vectors and much cheaper than `full_matrices=True`); revisit `torch.svd_lowrank` only if decomposition cost becomes a concern.
- **Device/dtype:** the returned `nn.Linear` inherits the source layer's device and dtype. The SVD is computed on the source device, promoting to float32 when the weight is lower precision (fp16/bf16) for numerical stability, then casting the truncated weight/bias back to the original dtype.
- **Contract edge cases:** `k == out` returns an identity-equivalent projection; a `None` bias passes through as `None`. Stated so #11 can rely on the behaviour.

## References

1. https://doi.org/10.48550/arXiv.2604.02659
2. https://doi.org/10.1109/ICMLT65785.2025.11193154
3. https://doi.org/10.1109/ICMEW.2019.00073

**Comment by markvilar** [comment]
Here is a reference to another repository where we have added SVD-truncation of the final linear layer of megaloc:
https://github.com/markvilar/benthloc-experiments/blob/dev/src/experiments/kappaplace/kappaforge/megaloc_builder.py
